src/containers/TransactionHistoryModal/TransactionHistoryModal.js

```diff
@@ -0,0 +1,93 @@
+import React, { useState, useEffect } from 'react';
+import { DataGrid } from '@material-ui/data-grid';
+import { useHistory, useRouteMatch } from 'react-router-dom';
+import { DateTime } from 'luxon';
+import { getUserById } from '../../services/databaseService';
+import Modal from '../../components/Modal/Modal';
+import { exchangeRoute } from '../../utils/routes';
+import useUserId from '../../hooks/useUserId';
+
+const columns = [
+  {
+    field: 'id',
+    headerName: 'ID',
+    minWidth: 90
+  },
+  {
+    field: 'currency',
+    headerName: 'Currency',
+    minWidth: 140
+  },
+  {
+    field: 'amount',
+    headerName: 'Amount',
+    type: 'number',
+    minWidth: 130
+  },
+  {
+    field: 'price',
+    headerName: 'Price',
+    minWidth: 110
+  },
+  {
+    field: 'action',
+    headerName: 'Action',
+    minWidth: 120
+  },
+  {
+    field: 'time',
+    headerName: 'Time',
+    minWidth: 225
+  }
+];
+
+const mapTransactionsToGridData = (transactions) => {
+  let id = 1;
+  return transactions.map(({ time, transactionInfo, error }) => {
+    const { currency, amount, price, isBuying } = transactionInfo;
+    const getAction = () => {
+      const action = isBuying ? 'Buy' : 'Sell';
+      return error ? `${action} (Failed)` : action;
+    };
+    return {
+      id: id++,
+      currency,
+      amount,
+      price: price.toFixed(4),
+      action: getAction(),
+      time: DateTime.fromISO(time).toHTTP()
+    };
+  });
+};
+
+const TransactionHistoryModal = () => {
+  const history = useHistory();
+  const [transactions, setTransactions] = useState([]);
+  const { params: { currency } } = useRouteMatch();
+  const [userId] = useUserId();
+
+  useEffect(() => {
+    const setUserTransactions = async (userId) => {
+      setTransactions((await getUserById(userId)).transactions);
+    };
+    setUserTransactions(userId);
+  }, []);
```

Comment:
I would suggest to keep transactions in redux store globally.

Reply:
Lets describe actual flow:
1. user logs in => database service fetch data and updates redux store => component receive data from store, doesn't have its own state
2. in case of new transaction appear => middleware updates database using service => redux updates it's own state

So after all database service is performing GET only ONCE (after "login"). Performing PATCH on given list of actions. Redux state updates on every action.